iroha/src/event.rs

```diff
@@ -64,6 +68,115 @@ impl Into<Event> for Occurrence {
     }
 }
 
+/// Module `connection` provides functionality needed for Iroha Events consumers.
+pub mod connection {
+    use super::*;
+
+    /// Criteria to filter `Occurrences` based on.
+    #[derive(Clone, Debug, Io, Encode, Decode)]
+    pub struct Criteria {
+        occurrence_type: OccurrenceType,
+        entity_type: EntityType,
+    }
+
+    /// Which type of `Occurrences` should be added to filter.
+    #[derive(Clone, Debug, Io, Encode, Decode)]
+    pub enum OccurrenceType {
+        /// Filter `Occurrence::Created` events.
+        Created,
+        /// Filter `Occurrence::Updated` events.
+        Updated,
+        /// Filter `Occurrence::Deleted` events.
+        Deleted,
+        /// Filter all types of `Occurrence`.
+        All,
+    }
+
+    /// Which type of `Entities` should be added to filter.
+    #[derive(Clone, Debug, Io, Encode, Decode)]
+    pub enum EntityType {
+        /// Filter `Entity::Account` events.
+        Account,
+        /// Filter `Entity::AssetDefinition` events.
+        AssetDefinition,
+        /// Filter `Entity::Asset` events.
+        Asset,
+        /// Filter `Entity::Domain` events.
+        Domain,
+        /// Filter `Entity::Peer` events.
+        Peer,
+        /// Filter `Entity::Transaction` events.
+        Transaction,
```

Comment:
I think for transaction we also need to track when they are deleted, so it will be good to add a `Deleted` event when they are dropped in a Queue due to TTL.

Reply:
For transaction validation when they are dropped in Sumeragi at least on the step by the leader would also be good. But maybe it's a thing for a separate PR.

Reply:
I mean for the app developers, as discussed in this [RFC](https://wiki.hyperledger.org/display/iroha/Prevent+replay+of+rejected+transactions) it would be a good solution.

Reply:
We can't "delete" something from blockchain so all this types of events are about "Updated".

Reply:
I mean you have `OccurenceType::Deleted`. And yes ones something in a blockchain it can not be deleted. But I mean some transactions are rejected because of TTL or being invalid. And as we discussed in RFC, the app developers might want to know if the transaction was rejected.

Reply:
It is possible to create a separate OccurenceType for that like `Rejected` but then it would not really fit other entity types.

Reply:
I get it - this is top level type applicable to all entities. Maybe we will make it more fine grained by additional parameters, or will rework this. Will see in next PR with real examples of senders and tests for consumers.